src/chain.jl

```diff
@@ -0,0 +1,49 @@
+
+import Flux: ChainRulesCore
+# Some experiments with chain to start removing the need for recur to be mutable.
+# As per the conversation in the recurrent network rework issue.
+
+# Main difference between this and the _applychain function is we return a new chain
+# with the internal state modified as well as the output of applying x to the chain.
+apply(chain::Flux.Chain, x) = begin
+  layers, out = _apply(chain.layers, x)
+  Flux.Chain(layers), out
+end
+
+_apply(layers::NamedTuple{NMS, TPS}, x) where {NMS, TPS} = begin
+  layers, out = _apply(Tuple(layers), x)
+  NamedTuple{NMS}(layers), out
+end
+
+function _scan(layers::AbstractVector, x)
+  new_layers = typeof(layers)(undef, length(layers))
+  for (idx, f) in enumerate(layers)
+    new_layers[idx], x = _apply_to_layer(f, x)
+  end
+  new_layers, x
+end
+
+function ChainRulesCore.rrule(::typeof(_scan), layers, x)
+  function _scan_pullback(dy)
+    throw("_scan Pullback not implemented")
+    return (ChainRulesCore.NoTangent(), ChainRulesCore.NoTangent())
```

Comment:
```suggestion
    error("_scan Pullback not implemented")
```

Reply:
This should return `ChainRulesCore.NotImplemented`, but currently Zygote (incorrectly) ignores that.